P1684-mdarray/P1684.md

```diff
@@ -334,6 +342,397 @@ In R3 we decided to radically cut back on constructors and simply leave those ou
+
+---
+
+```cpp
+  template<class OtherElementType, class OtherExtents, 
+           class OtherLayoutPolicy, class OtherContainer>
+    explicit(@_see below_@)
+    constexpr mdarray(const mdarray<OtherElementType,
+                    OtherExtents,
+                    OtherLayoutPolicy,
+                    OtherContainer>& other);
+
+  template<class OtherElementType, class OtherExtents,
+           class OtherLayoutPolicy, class Accessor>
+    explicit(@_see below_@)
+    constexpr mdarray(const mdspan<OtherElementType,
+                                   OtherExtents,
+                                   OtherLayoutPolicy,
+                                   Accessor>& other);
+
+  template<class OtherElementType, class OtherExtents, 
+           class OtherLayoutPolicy, class OtherContainer, 
+           class Alloc>
+    explicit(@_see below_@)
+    constexpr mdarray(const mdarray<OtherElementType,
+                                    OtherExtents, 
+                                    OtherLayoutPolicy,
+                                    OtherContainer>& other,
+                      const Alloc& a);
+
+  template<class OtherElementType, class OtherExtents,
+           class OtherLayoutPolicy, class Accessor,
+           class Alloc>
+    explicit(@_see below_@)
+    constexpr mdarray(const mdspan<OtherElementType,
+                                   OtherExtents,
+                                   OtherLayoutPolicy,
+                                   Accessor>& other,
+                      const Alloc& a);
+```
+
+```cpp
+  template<class OtherElementType, class OtherExtents, 
+           class OtherLayoutPolicy, class OtherContainer, 
+           class... Args>
+    explicit(@_see below_@)
+    constexpr mdarray(const mdarray<OtherElementType,
+                                    OtherExtents, 
+                                    OtherLayoutPolicy,
+                                    OtherContainer>& other,
+                      Args&&... args);
+
+  template<class OtherElementType, class OtherExtents,
+           class OtherLayoutPolicy, class Accessor,
+           class... Args>
+    explicit(@_see below_@)
+    constexpr mdarray(const mdspan<OtherElementType,
+                                   OtherExtents,
+                                   OtherLayoutPolicy,
+                                   Accessor>& other,
+                      Args&&... args);
+```
+
+---
+
+```cpp
+template <class Container, class... Integrals>
+requires((is_convertible_v<Integrals, size_t> && ...) &&
+         sizeof...(Integrals) > 0)
+explicit mdarray(const Container&, Integrals...)
+  -> mdarray<typename Container::value_type,
+             dextents<size_t, sizeof...(Integrals)>,
+             layout_right,
+             Container>;
+
+template<class Container, class Extents>
+mdarray(const Container&, const Extents&)
+  -> mdarray<typename Container::value_type,
+             Extents,
+             layout_right,
+             Container>;
```

Comment:
Do we think these deduction guides will work in practice?  Would we need to "constrain" `Extents` here?

Reply:
I think the `requires` clause at line 584 is sufficient to differentiate it from the other guides.

I don't know that it works in practice.  These two are identical CTAD constructors which map to two different types:

```c++
template<class Container, class Extents>
mdarray(Container&&, const Extents&)
  -> mdarray<typename Container::value_type, Extents, layout_right, Container>;

template<class Container, class Mapping>
mdarray(Container&&, const Mapping&)
  -> mdarray<typename Container::value_type, typename Mapping::extents_type, layout_right, Container>;
```

Another way to write the part before the `->` is:
```c++
template<class A, class B>
mdarray(A&&, const B&)
```

We'd be counting on `typename Mapping::extents_type` being SFINAEd out when `Mapping == std::extents<...>` AND being a better match when it is a `layout_*::mapping<...>`.  I don't think that works (but I could be mistaken).

Reply:
@nliber wrote:
> I don't know that it works in practice. 

Would you consider writing a brief test?

Reply:
Let me try a simpler explanation.

Every function to the left of the `->` in the set of explicit deduction guides are functions added to the overload set when using `mdarray` without any angle brackets.  (There are also implicit deduction guides which also adds to the overload set.). Like every other kind of function in C++, one cannot have duplicates in the overload set.

Simple example:
```c++
template<typename T>
struct A {
    A(T&&) {}
};

template<typename T>
A(T&&)->A<T>;

template<typename T>
A(T&&)->A<T>;
```
[Compiler Explorer](https://godbolt.org/z/1Yah1KMaT) says:

```
<source>:10:1: error: deduction guide 'template<class T> A(T&&)-> A<T>' redeclared
   10 | A(T&&)->A<T>;
      | ^
<source>:7:1: note: 'template<class T> A(T&&)-> A<T>' previously declared here
    7 | A(T&&)->A<T>;
      | ^
Compiler returned: 1
```

Reply:
@nliber Would you consider adding this argument to the paper?

I'm hoping we can avoid the mdspan experience of trying to mash a bunch of CTAD fixes in near the end of the process : - ) .